File: 4Feb2024.py
class Solution:
    def idx(self,s : str,i: int) :
        if(ord(s[i])>=ord('a')) :
            return ord(s[i])-ord('a')+26
        return ord(s[i])-ord('A')
    
    def minWindow(self, s: str, t: str) -> str:
        fre = [0]*52
        for i in range(len(t)) :
            fre[self.idx(t,i)]-=1
        strt=0
        end=0
        i,j,ans=0,0,inf
        while(j<len(s)) :
            isValid=True
            for f in fre :
                if(f<0) :
                    isValid=False
            if(isValid) :
                fre[self.idx(s,i)]-=1
                if(ans>j-i) :
                    ans=j-i
                    strt=i
                    end=j
                i+=1
            else :
                fre[self.idx(s,j)]+=1
                j+=1
        
        while(i<len(s)) :
            isValid=True
            for f in fre :
                if(f<0) :
                    isValid=False
            if(isValid) :
                fre[self.idx(s,i)]-=1
                if(ans>j-i) :
                    ans=j-i
                    strt=i
                    end=j
                i+=1
            else :
                break
        return s[strt:end]

File: 4Feb2024.java
class Solution {
    public int idx(char c) {
        if (c >= 'a') {
            return c - 'a' + 26;
        }
        return c - 'A';
    }

    public String minWindow(String s, String t) {
        int[] fre = new int[52];
        for (int i = 0; i < t.length(); i++)
            fre[idx(t.charAt(i))]--;
        int strt = 0, end = 0;
        int i = 0, j = 0, ans = Integer.MAX_VALUE;
        while (j < s.length()) {
            boolean isValid = true;
            for (int f : fre)
                if (f < 0)
                    isValid = false;
            // System.out.println(i+" "+j+" "+Arrays.toString(fre));
            if (isValid) {
                fre[idx(s.charAt(i))]--;
                if (ans > j - i) {
                    ans = j - i;
                    strt = i;
                    end = j;
                }
                i++;
            } else {
                fre[idx(s.charAt(j))]++;
                j++;
            }
        }

        while (i < s.length()) {
            boolean isValid = true;
            for (int f : fre)
                if (f < 0)
                    isValid = false;
            // System.out.println(i+" "+j+" "+Arrays.toString(fre));
            if (isValid) {
                fre[idx(s.charAt(i))]--;
                if (ans > j - i) {
                    ans = j - i;
                    strt = i;
                    end = j;
                }
                i++;
            } else {
                break;
            }
        }
        // System.out.println(ans);
        return s.substring(strt, end);
    }
}